lib/srv/db/cassandra/engine.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cassandra
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/hex"
+	"errors"
+	"io"
+	"net"
+
+	"github.com/datastax/go-cassandra-native-protocol/client"
+	"github.com/datastax/go-cassandra-native-protocol/message"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/defaults"
+	libevents "github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/srv/db/cassandra/protocol"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/common/role"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func init() {
+	common.RegisterEngine(newEngine, defaults.ProtocolCassandra)
+}
+
+// newEngine create new Cassandra engine.
+func newEngine(ec common.EngineConfig) common.Engine {
+	return &Engine{
+		EngineConfig: ec,
+	}
+}
+
+// Engine implements common.Engine.
+type Engine struct {
+	// EngineConfig is the common database engine configuration.
+	common.EngineConfig
+	// clientConn is a client connection.
+	clientConn *protocol.Conn
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+	// handshakeTriggered is set to true if handshake was triggered and
+	// used to indicated that custom errors should be sent to the client.
+	// Cassandra wire protocol relies on streamID to that needs to match the request value
+	// so sending a custom error to the client requires to read a first message send by the client.
+	handshakeTriggered bool
+}
+
+// SendError send a Cassandra ServerError to  error to the client if handshake was not yet initialized by the client.
+// Cassandra wire protocol relies on streamID to that are set by the client and server response needs to
+// set the correct streamID in order to get streamID SendError reads a first message send by the client.
+func (e *Engine) SendError(sErr error) {
+	if utils.IsOKNetworkError(sErr) || sErr == nil {
+		return
+	}
+	e.Log.Debugf("Cassandra connection error: %v.", sErr)
+	if e.handshakeTriggered {
+		return
```

Comment:
Can you add a code comment justifying ignoring an error after triggering the handshake?